### PR TITLE
Disjoint Union DLC

### PR DESCRIPTION
**Pull request by nkohen**
Implemented disjoint union DLC related data structures. Also refactored dlcTest to have a `DLCTest` trait with all of the constructors and parameter case classes and broke up `DLCClientTest` into separate test files for each kind of DLC. Tests in dlcTest now pass for me in under 40 seconds!

`dlcWallet` will need some restructuring work (@benthecarman).

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 1]
```diff
@@ -127,167 +263,149 @@ case class ContractInfo(
...
       totalCollateral,
       ContractOraclePair.fromDescriptorOracle(contractDescriptor, oracleInfo))
   }
 }
+
+case class DisjointUnionContractInfo(contracts: Vector[SingleContractInfo])
+    extends ContractInfo
+    with TLVSerializable[ContractInfoV1TLV] {
+
+  override val totalCollateral: Satoshis = contracts.head.totalCollateral
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala]: nit: should add a `require(contracts.nonEmpty)` so a better error message is given

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 2]
```diff
@@ -127,167 +263,149 @@ case class ContractInfo(
...
+
+  override def toTLV: ContractInfoV1TLV = {
+    ContractInfoV1TLV(
+      totalCollateral,
+      contracts.map(contract =>
+        (contract.contractDescriptor.toTLV, contract.oracleInfo.toTLV)))
+  }
+
+  /** @inheritdoc */
+  override val max: Satoshis = contracts.map(_.max).maxBy(_.toLong)
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala]: You should be able to use `.max` instead of `.maxBy`

**Review by benthecarman — COMMENTED** [review]
couple nits

**Comment by Christewart** [comment]
needs rebase

**Comment by Christewart** [comment]
superseded #3839

**Pull request closed without merging**